src/kaleidoscope/hid.cpp

```diff
@@ -51,42 +52,172 @@
+
+// This is used to determine which modifier flags will be allowed to be added to the
+// current keyboard HID report. It gets set during any cycle where one or more new keys
+// is toggled on and presists until the next cycle with a newly detected keypress.
+
+static uint8_t modifier_flag_mask{0};
+
+
+
+// The functions in this namespace are primarily to solve the problem of rollover from a
+// key with a modifier flag (e.g. `LSHIFT(Key_T)`) to one without (e.g. `Key_H`), which
+// used to result in the mod flag being applied to keys other than the one with the
+// flag. By using `modifier_flag_mask`, we can mask out any modifier flags
+// that aren't attached to modifier keys or keys pressed in the most recent cycle,
+// mitigatingthe rollover problem, and getting the intended `The` instead of `THe`.
+
+// requested_modifiers is bitmask of the modifiers attached to any non-modifier key found to be pressed
+// during the most recent cycle. For example, it would include modifiers attached to Key_A, but not 
+// modifiers attached to Key_LeftControl)
+
+static uint8_t requested_modifiers{0};
+
+// last_keycode_toggled_on is the keycode of the key most recently toggled on
+// for this report.  This is set when the key is pressed and cleared after
+// the report is sent.
+
+static uint8_t last_keycode_toggled_on{0};
+
+
+void resetModifierTracking(void) {
+  last_keycode_toggled_on = 0;
+  requested_modifiers = 0;
 }
 
-void pressRawKey(Key mappedKey) {
-  WITH_BOOTKEYBOARD_PROTOCOL {
-    BootKeyboard.press(mappedKey.keyCode);
-    return;
-  }
+// This returns true if the key is a keyboard key, and its keycode is a modifier.
+bool isModifierKey(Key mappedKey) {
+  // If it's not a keyboard key, return false
+  if (mappedKey.flags & (SYNTHETIC | RESERVED))
+    return false;
 
-  Keyboard.press(mappedKey.keyCode);
+  return (mappedKey.keyCode >= HID_KEYBOARD_FIRST_MODIFIER &&
+          mappedKey.keyCode <= HID_KEYBOARD_LAST_MODIFIER);
 }
 
-void pressKey(Key mappedKey) {
-  if (mappedKey.flags & SHIFT_HELD) {
+// requestModifiers takes a bitmap of modifiers that might apply
+// to the next USB HID report and adds them to a bitmap of all such modifiers.
+
+void requestModifiers(byte flags) {
+  requested_modifiers |= flags;
+}
+
+
+// pressModifiers takes a bitmap of modifier keys that must be included in
+// the upcoming USB HID report and passes them through to KeyboardioHID
+// immediately
+void pressModifiers(byte flags) {
+  if (flags & SHIFT_HELD) {
     pressRawKey(Key_LeftShift);
   }
-  if (mappedKey.flags & CTRL_HELD) {
+  if (flags & CTRL_HELD) {
     pressRawKey(Key_LeftControl);
   }
-  if (mappedKey.flags & LALT_HELD) {
+  if (flags & LALT_HELD) {
     pressRawKey(Key_LeftAlt);
   }
-  if (mappedKey.flags & RALT_HELD) {
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}
```

Comment:
I had a comment here indicating that this was the end of the anonymous namespace. I find that very useful to keep from having to search for the matching curly brace.

Reply:
Indeed, I remove it because I really, really dislike block closing comments. In my experience, they tend to get out of sync and become lies quite quickly.

That said, I went to the style guide to get myself some backup. At least as of now, I don't have that backup ;) The style guide inherited from Google doesn't agree with me. I'll put it back.

Reply:
I've started to switch from `astyle` to `clang-format`, and one of the things it does (at least with the Google style option) is add (and rewrite) comments that close namespaces, so if it gets run before commits, it would prevent those comments from getting stale.